airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py

```diff
@@ -194,6 +197,91 @@ class ShoppingPerformanceReport(IncrementalGoogleAdsStream):
+        If you need to enable it, uncomment the next line instead of `return None` and modify your config
+        """
+        # return self.custom_query_config.get("primary_key") or None
+        return None
+
+    @property
+    def name(self):
+        return self.custom_query_config["table_name"]
+
+    @property
+    def cursor_field(self) -> str:
+        """
+        The incremental is disabled. Config / spec should not provide the cursor_field.
+        It will be ignored if provided.
+        However, this return should be kept for case we wanna support it.
+        Disabled cursor_field should be always empty array or string, to keep the internal logic
+            (get length of cursor_field).
+        Since it is not provided, the stream will be full refresh anyway.
+        The inheritance from the Incremental stream is made for supporting both types,
+            and need to be kept.
+        If you need to enable this option, uncomment the first return and modify your config
+        """
+        # return self.custom_query_config.get("cursor_field") or []
+        return []
+
+    def stream_slices(self, stream_state: Mapping[str, Any] = None, **kwargs) -> Iterable[Optional[Mapping[str, any]]]:
+        if not self.cursor_field:
+            return [None]
+        return super().stream_slices(stream_state=stream_state, **kwargs)
+
+    def get_query(self, stream_slice: Mapping[str, Any] = None) -> str:
+        if not self.cursor_field:
+            return self.user_defined_query
+        start_date, end_date = self.get_date_params(stream_slice, self.cursor_field)
+        final_query = (
+            self.user_defined_query
+            + f"\nWHERE {self.cursor_field} > '{start_date}' AND {self.cursor_field} < '{end_date}' ORDER BY {self.cursor_field} ASC"
+        )
+        return final_query
+
+    def get_json_schema(self):
+        local_json_schema = {
+            "$schema": "http://json-schema.org/draft-07/schema#",
+            "type": "object",
+            "properties": {},
+            "additionalProperties": True,
+        }
+        # full list {'ENUM', 'STRING', 'DATE', 'DOUBLE', 'RESOURCE_NAME', 'INT32', 'INT64', 'BOOLEAN', 'MESSAGE'}
+        google_datatype_mapping = {"INT64": "integer", "DOUBLE": "number", "STRING": "string", "BOOLEAN": "boolean", "DATE": "string"}
+        fields = self.user_defined_query.lower().split("select")[1].split("from")[0].strip()
+        google_resource_name = self.user_defined_query.lower().split("from", 1)[1].strip()
+        google_resource_name = re.split("\\s+", google_resource_name)[0]
+
+        fields = fields.split(",")
+        fields = [i.strip() for i in fields]
+        fields = list(dict.fromkeys(fields))
+
+        google_schema = ResourceSchemaLoader(package_name_from_class(self.__class__)).get_schema(f"shared/v8_{google_resource_name}")
+        for field in fields:
+            node = google_schema.get("fields").get(field).get("field_details")
```

Comment:
can you add unit tests to test the following: 

1. Correct parsing of the query
2. correct conversion from google type to airbyte type